src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+            SymbolToken[] annotationArray = new SymbolToken[annotationSids.size()];
+            for (int i = 0; i < annotationArray.length; i++) {
+                annotationArray[i] = getSymbolToken(annotationSids.get(i));
+            }
+            return annotationArray;
+        }
+        return SymbolToken.EMPTY_ARRAY;
+    }
+
+    private static final Iterator<String> EMPTY_ITERATOR = new Iterator<String>() {
+
+        @Override
+        public boolean hasNext() {
+            return false;
+        }
+
+        @Override
+        public String next() {
+            return null;
+        }
+
+        @Override
+        public void remove() {
+            throw new UnsupportedOperationException("Cannot remove from an empty iterator.");
+        }
+    };
+
+    @Override
+    public Iterator<String> iterateTypeAnnotations() {
+        if (hasAnnotations) {
+            annotationIterator.reset();
+            return annotationIterator;
+        }
+        return EMPTY_ITERATOR;
+    }
+
+    @Override
+    public int getFieldId() {
+        return fieldNameSid;
+    }
+
+    @Override
+    public String getFieldName() {
+        if (fieldNameSid < 0) {
+            return null;
+        }
+        String fieldName = getSymbol(fieldNameSid);
+        if (fieldName == null) {
+            throw new UnknownSymbolException(fieldNameSid);
+        }
+        return fieldName;
+    }
+
+    @Override
+    public SymbolToken getFieldNameSymbol() {
+        if (fieldNameSid < 0) {
+            return null;
+        }
+        return getSymbolToken(fieldNameSid);
+    }
+
+    @Override
+    public boolean isNullValue() {
+        return valueTypeID != null && valueTypeID.isNull;
+    }
+
+    @Override
+    public boolean isInStruct() {
+        return !containerStack.isEmpty() && containerStack.peek().type == IonType.STRUCT;
+    }
+
+    @Override
+    public boolean booleanValue() {
+        requireType(IonType.BOOL);
+        return valueTypeID.lowerNibble == 1;
+    }
+
+    @Override
+    public <T> T asFacet(Class<T> facetType) {
+        return null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        // NOTE: If we want to replace the other binary IonReader implementation with this one, the following
+        // validation could be performed in next() if incremental mode is not enabled. That would allow this
+        // implementation to behave in the same way as the other implementation when an incomplete value is
+        // encountered.
+        if (lookahead.isSkippingCurrentValue()) {
+            throw new IonException("Unexpected EOF.");
+        }
+        if (lookahead.available() > 0 && lookahead.moreDataRequired()) {
+            if (lookahead.getIvmIndex() < 0
+                || lookahead.available() != _Private_IonConstants.BINARY_VERSION_MARKER_SIZE) {
+                throw new IonException("Unexpected EOF.");
+            }
+        }
```

Comment:
Could extract a method to move one step closer. Your mileage may vary, YAGNI, etc.

Reply:
True. I'm inclined to wait until we know we're going to need it.